File: node/src/protocol.js
import { createHash, randomBytes } from "node:crypto";
import { readFileSync } from "node:fs";
import path from "node:path";

import mimeTypes from "mime-types";

export const PROTOCOL_VERSION = "QRB1";
export const MAX_FILE_SIZE = 5 * 1024 * 1024;
export const MANIFEST_INTERVAL = 20;

export const PROFILES = Object.freeze({
  safe: Object.freeze({
    name: "safe",
    chunkSize: 480,
    fps: 6,
    errorCorrection: "M",
  }),
  fast: Object.freeze({
    name: "fast",
    chunkSize: 900,
    fps: 10,
    errorCorrection: "L",
  }),
});

const CRC32_TABLE = new Uint32Array(256);
for (let index = 0; index < CRC32_TABLE.length; index += 1) {
  let value = index;
  for (let bit = 0; bit < 8; bit += 1) {
    value = value & 1 ? 0xedb88320 ^ (value >>> 1) : value >>> 1;
  }
  CRC32_TABLE[index] = value >>> 0;
}

export function b64urlEncode(data) {
  return Buffer.from(data).toString("base64url");
}

export function b64urlDecode(value) {
  if (!/^[A-Za-z0-9_-]*$/.test(value) || value.length % 4 === 1) {
    throw new Error("invalid base64url payload");
  }
  return Buffer.from(value, "base64url");
}

export function crc32Hex(data) {
  let crc = 0xffffffff;
  for (const byte of Buffer.from(data)) {
    crc = CRC32_TABLE[(crc ^ byte) & 0xff] ^ (crc >>> 8);
  }
  return ((crc ^ 0xffffffff) >>> 0).toString(16).padStart(8, "0");
}

export function sanitizeFilename(value) {
  const portablePath = String(value).replaceAll("\\", "/");
  let name = path.posix.basename(portablePath);
  name = name
    .replace(/[\x00-\x1f\x7f/\\:*?"<>|]/g, "_")
    .replace(/^[ .]+|[ .]+$/g, "");
  name = [...name].slice(0, 180).join("");
  return name || "received-file";
}

function manifestFrame(sessionId, manifest) {
  const raw = Buffer.from(JSON.stringify(manifest), "utf8");
  return `${PROTOCOL_VERSION}|M|${sessionId}|${b64urlEncode(raw)}|${crc32Hex(
    raw
  )}`;
}

function dataFrame(sessionId, index, total, chunk) {
  return `${PROTOCOL_VERSION}|D|${sessionId}|${index}|${total}|${b64urlEncode(
    chunk
  )}|${crc32Hex(chunk)}`;
}

export function parseFrame(value) {
  const parts = value.split("|");
  if (parts.length < 5 || parts[0] !== PROTOCOL_VERSION) {
    throw new Error("not a QRB1 frame");
  }
  const kind = parts[1];
  const sessionId = parts[2];
  if (!/^[0-9a-f]{16}$/.test(sessionId)) {
    throw new Error("invalid session id");
  }

  let payload;
  let crc32;
  let index = null;
  let total = null;
  if (kind === "M" && parts.length === 5) {
    payload = b64urlDecode(parts[3]);
    crc32 = parts[4];
  } else if (kind === "D" && parts.length === 7) {
    if (!/^\d+$/.test(parts[3]) || !/^\d+$/.test(parts[4])) {
      throw new Error("invalid chunk position");
    }
    index = Number(parts[3]);
    total = Number(parts[4]);
    if (
      !Number.isSafeInteger(index) ||
      !Number.isSafeInteger(total) ||
      index < 0 ||
      total < 1 ||
      index >= total
    ) {
      throw new Error("chunk position out of range");
    }
    payload = b64urlDecode(parts[5]);
    crc32 = parts[6];
  } else {
    throw new Error("invalid QRB1 frame shape");
  }

  if (!/^[0-9a-f]{8}$/.test(crc32)) {
    throw new Error("invalid crc32");
  }
  if (crc32Hex(payload) !== crc32) {
    throw new Error("crc32 mismatch");
  }
  return {
    version: PROTOCOL_VERSION,
    kind,
    sessionId,
    payload,
    crc32,
    index,
    total,
  };
}

export class TransferSession {
  constructor(filePath, profile, options = {}) {
    if (!profile || !Object.values(PROFILES).includes(profile)) {
      throw new Error("invalid transfer profile");
    }
    this.filePath = filePath;
    this.profile = profile;
    this.data = readFileSync(filePath);
    if (this.data.length > MAX_FILE_SIZE) {
      throw new Error(`file exceeds ${MAX_FILE_SIZE} byte limit`);
    }

    this.sessionId = options.sessionId ?? randomBytes(8).toString("hex");
    if (!/^[0-9a-f]{16}$/.test(this.sessionId)) {
      throw new Error("sessionId must be 16 lowercase hex characters");
    }

    this.chunks = [];
    for (
      let offset = 0;
      offset < this.data.length;
      offset += profile.chunkSize
    ) {
      this.chunks.push(this.data.subarray(offset, offset + profile.chunkSize));
    }
    const fileName = sanitizeFilename(filePath);
    this.manifest = {
      chunkSize: profile.chunkSize,
      fileName,
      fileSize: this.data.length,
      mime: mimeTypes.lookup(fileName) || "application/octet-stream",
      protocol: PROTOCOL_VERSION,
      sha256: createHash("sha256").update(this.data).digest("hex"),
      totalChunks: this.chunks.length,
    };
    this.manifestFrame = manifestFrame(this.sessionId, this.manifest);
    this.dataFrames = this.chunks.map((chunk, index) =>
      dataFrame(this.sessionId, index, this.chunks.length, chunk)
    );
    this.frames = this.buildCycle();
  }

  buildCycle() {
    if (this.dataFrames.length === 0) {
      return [this.manifestFrame];
    }
    const cycle = [this.manifestFrame];
    this.dataFrames.forEach((frame, index) => {
      cycle.push(frame);
      const sent = index + 1;
      if (sent % MANIFEST_INTERVAL === 0 && sent < this.dataFrames.length) {
        cycle.push(this.manifestFrame);
      }
    });
    return cycle;
  }

  frameAt(position) {
    if (!Number.isSafeInteger(position) || position < 0) {
      throw new Error("position must be a non-negative integer");
    }
    return this.frames[position % this.frames.length];
  }

  get cycleLength() {
    return this.frames.length;
  }
}

File: src/qrbeam/protocol.py
from __future__ import annotations

import base64
import hashlib
import json
import mimetypes
import re
import secrets
import zlib
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

PROTOCOL_VERSION = "QRB1"
MAX_FILE_SIZE = 5 * 1024 * 1024
MANIFEST_INTERVAL = 20


@dataclass(frozen=True)
class Profile:
    name: str
    chunk_size: int
    fps: int
    error_correction: Literal["L", "M"]


PROFILES: dict[str, Profile] = {
    "safe": Profile("safe", chunk_size=480, fps=6, error_correction="M"),
    "fast": Profile("fast", chunk_size=900, fps=10, error_correction="L"),
}


@dataclass(frozen=True)
class ParsedFrame:
    version: str
    kind: Literal["M", "D"]
    session_id: str
    payload: bytes
    crc32: str
    index: int | None = None
    total: int | None = None


def b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def b64url_decode(value: str) -> bytes:
    padding = "=" * (-len(value) % 4)
    return base64.urlsafe_b64decode(value + padding)


def crc32_hex(data: bytes) -> str:
    return f"{zlib.crc32(data) & 0xFFFFFFFF:08x}"


def sanitize_filename(value: str) -> str:
    name = Path(value).name
    name = re.sub(r"[\x00-\x1f\x7f/\\:*?\"<>|]", "_", name)
    name = name.strip(" .")
    return name[:180] or "received-file"


def _manifest_frame(session_id: str, manifest: dict[str, Any]) -> str:
    raw = json.dumps(
        manifest,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")
    return f"{PROTOCOL_VERSION}|M|{session_id}|{b64url_encode(raw)}|{crc32_hex(raw)}"


def _data_frame(session_id: str, index: int, total: int, chunk: bytes) -> str:
    return (
        f"{PROTOCOL_VERSION}|D|{session_id}|{index}|{total}|"
        f"{b64url_encode(chunk)}|{crc32_hex(chunk)}"
    )


def parse_frame(value: str) -> ParsedFrame:
    parts = value.split("|")
    if len(parts) < 5 or parts[0] != PROTOCOL_VERSION:
        raise ValueError("not a QRB1 frame")
    kind = parts[1]
    session_id = parts[2]
    if not re.fullmatch(r"[0-9a-f]{16}", session_id):
        raise ValueError("invalid session id")

    if kind == "M" and len(parts) == 5:
        payload = b64url_decode(parts[3])
        crc = parts[4]
        frame = ParsedFrame(PROTOCOL_VERSION, "M", session_id, payload, crc)
    elif kind == "D" and len(parts) == 7:
        try:
            index = int(parts[3])
            total = int(parts[4])
        except ValueError as exc:
            raise ValueError("invalid chunk position") from exc
        if index < 0 or total < 1 or index >= total:
            raise ValueError("chunk position out of range")
        payload = b64url_decode(parts[5])
        crc = parts[6]
        frame = ParsedFrame(
            PROTOCOL_VERSION,
            "D",
            session_id,
            payload,
            crc,
            index=index,
            total=total,
        )
    else:
        raise ValueError("invalid QRB1 frame shape")

    if not re.fullmatch(r"[0-9a-f]{8}", frame.crc32):
        raise ValueError("invalid crc32")
    if crc32_hex(frame.payload) != frame.crc32:
        raise ValueError("crc32 mismatch")
    return frame


class TransferSession:
    def __init__(
        self,
        file_path: Path,
        profile: Profile,
        *,
        session_id: str | None = None,
    ) -> None:
        self.file_path = Path(file_path)
        self.profile = profile
        self.data = self.file_path.read_bytes()
        if len(self.data) > MAX_FILE_SIZE:
            raise ValueError(f"file exceeds {MAX_FILE_SIZE} byte limit")
        self.session_id = session_id or secrets.token_hex(8)
        if not re.fullmatch(r"[0-9a-f]{16}", self.session_id):
            raise ValueError("session_id must be 16 lowercase hex characters")

        self.chunks = [
            self.data[offset : offset + profile.chunk_size]
            for offset in range(0, len(self.data), profile.chunk_size)
        ]
        mime = mimetypes.guess_type(self.file_path.name)[0] or "application/octet-stream"
        self.manifest: dict[str, Any] = {
            "chunkSize": profile.chunk_size,
            "fileName": sanitize_filename(self.file_path.name),
            "fileSize": len(self.data),
            "mime": mime,
            "protocol": PROTOCOL_VERSION,
            "sha256": hashlib.sha256(self.data).hexdigest(),
            "totalChunks": len(self.chunks),
        }
        self.manifest_frame = _manifest_frame(self.session_id, self.manifest)
        self.data_frames = [
            _data_frame(self.session_id, index, len(self.chunks), chunk)
            for index, chunk in enumerate(self.chunks)
        ]
        self.frames = self._build_cycle()

    def _build_cycle(self) -> list[str]:
        if not self.data_frames:
            return [self.manifest_frame]
        cycle = [self.manifest_frame]
        for index, frame in enumerate(self.data_frames, start=1):
            cycle.append(frame)
            if index % MANIFEST_INTERVAL == 0 and index < len(self.data_frames):
                cycle.append(self.manifest_frame)
        return cycle

    def frame_at(self, position: int) -> str:
        if position < 0:
            raise ValueError("position must be non-negative")
        return self.frames[position % len(self.frames)]

    @property
    def cycle_length(self) -> int:
        return len(self.frames)
